modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordBinaryManager.kt

```diff
@@ -0,0 +1,192 @@
+import java.net.URI
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+import java.net.URLEncoder
+import java.nio.file.Files
+import java.time.Duration
+
+/**
+ * For dynamically fetching and storing mirrord binary.
+ */
+object MirrordBinaryManager {
+    private const val cliBinary = "mirrord"
+    private const val versionEndpoint = "https://version.mirrord.dev/v1/version"
+    private const val downloadEndpoint = "https://github.com/metalbear-co/mirrord/releases/download"
+
+    /**
+     * @return version of mirrord binary
+     *
+     * @throws RuntimeException if `--version` command failed
+     */
+    private fun getVersion(cliPath: String): String {
+        val child = Runtime.getRuntime().exec(arrayOf(cliPath, "--version"))
+
+        val result = child.waitFor()
+        if (result != 0) {
+            MirrordLogger.logger.debug("`mirrord --version` failed with code $result")
+            throw RuntimeException("failed to get mirrord version")
+        }
+
+        return child.inputReader().readLine().split(' ')[1].trim()
+    }
+
+    /**
+     * @return path to the local installation of mirrord
+     */
+    private fun getLocalBinary(requiredVersion: String?): String? {
+        try {
+            val child = Runtime.getRuntime().exec(arrayOf("which", "mirrord"))
+            val result = child.waitFor()
+            if (result != 0) {
+                throw RuntimeException("`which` failed with code $result")
+            }
+            val foundInPath = child.inputReader().readLine().trim()
+
+            if (requiredVersion == null || requiredVersion == this.getVersion(foundInPath)) {
+                return foundInPath
+            }
+        } catch (e: Exception) {
+            MirrordLogger.logger.debug("failed to find mirrord in path", e)
+        }
+
+        try {
+            MirrordPathManager.getBinary(this.cliBinary, true)?.let {
+                if (requiredVersion == null || requiredVersion == this.getVersion(it)) {
+                    return it
+                }
+            }
+        } catch (e: Exception) {
+            MirrordLogger.logger.debug("failed to find mirrord in plugin storage", e)
+        }
+
+        return null
+    }
+
+    private fun getLatestSupportedVersion(product: String, timeout: Duration, project: Project): String {
+        val environment = CompletableFuture<String>()
+        val versionCheckTask = object : Task.Backgroundable(project, "mirrord", true) {
+            override fun run(indicator: ProgressIndicator) {
+                indicator.text = "mirrord is checking the latest supported version..."
+
+                val version = if (System.getenv("CI_BUILD_PLUGIN") == "true") {
+                    "test"
+                } else {
+                    MirrordVersionCheck.version ?: "unknown"
+                }
+
+                val url = StringBuilder(versionEndpoint)
+                        .append("?source=1")
```

Comment:
this could've been funny - source needs to be 3 ;)

Reply:
Done